In [1]:
library(tidyverse)
library(repr)
library(tidymodels)
source("cleanup.R")

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
── Attaching packages ────────────────────────────────────── tidymodels 1.1.1 ──

✔ broom        1.0.6     ✔ rsample      1.2.1
✔ dials        1.3.0     ✔ tune         1.1.2
✔ infer        1.0.7     ✔ workflows    1.1.4
✔ modeldata    1.4.0     ✔ workflowsets 1.0.1
✔ parsnip      1.2.1     ✔ yardstick    1.3.1
✔ recipes      1.1.0     

── Conflicts ───────────────────────────────────────── tidymodels_conflicts() ──
✖ scales::discard() masks purrr::discard()
✖ dplyr::filt

ERROR: Error in file(filename, "r", encoding = encoding): cannot open the connection


In [3]:
set.seed(1)

url_pl <- "https://raw.githubusercontent.com/takemil8088/ind-porject/refs/heads/main/players.csv"
players <- read_csv(url_pl) |>
select(experience,subscribe,played_hours,gender,Age) |>
filter(!is.na(experience),!is.na(subscribe),!is.na(played_hours),!is.na(gender),!is.na(Age)) |>
mutate(subscribe = as_factor(subscribe))
players

Rows: 196 Columns: 7
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): experience, hashedEmail, name, gender
dbl (2): played_hours, Age
lgl (1): subscribe

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


experience subscribe played_hours gender            Age
1   Pro        TRUE      30.3         Male               9 
2   Veteran    TRUE       3.8         Male              17 
3   Veteran    FALSE      0.0         Male              17 
4   Amateur    TRUE       0.7         Female            21 
5   Regular    TRUE       0.1         Male              21 
6   Amateur    TRUE       0.0         Female            17 
7   Regular    TRUE       0.0         Female            19 
8   Amateur    FALSE      0.0         Male              21 
9   Amateur    TRUE       0.1         Male              17 
10  Veteran    TRUE       0.0         Female            22 
11  Veteran    TRUE       1.6         Female            23 
12  Pro        TRUE       0.0         Male              17 
13  Beginner   TRUE       1.5         Female            25 
14  Amateur    TRUE       0.2         Male              22 
15  Veteran    TRUE       0.0         Non-binary        17 
16  Amateur    FALSE      0.0         Male              22 
17  Beginner   TRUE       0.0         Female            17 
18  Amateur    TRUE      48.4         Female            17 
19  Amateur    TRUE       0.5         Male              17 
20  Regular    TRUE       0.6         Male              19 
21  Regular    TRUE       0.3         Male               8 
22  Amateur    FALSE      0.1         Female            17 
23  Beginner   TRUE       1.0         Male              17 
24  Beginner   TRUE       0.0         Male              17 
25  Amateur    TRUE       0.7         Female            21 
26  Regular    TRUE       0.6         Female            28 
27  Regular    TRUE       0.0         Male              17 
28  Veteran    FALSE      0.0         Male              23 
29  Amateur    TRUE       1.8         Male              23 
30  Veteran    FALSE      0.1         Male              18 
⋮   ⋮          ⋮         ⋮            ⋮                 ⋮  
165 Regular    TRUE       0.1         Male              21 
166 Amateur    FALSE      1.0         Male              26 
167 Beginner   FALSE      0.3         Female            17 
168 Veteran    TRUE       0.1         Male              17 
169 Veteran    FALSE      0.1         Male              17 
170 Pro        FALSE      0.4         Male              21 
171 Beginner   FALSE      1.8         Male              32 
172 Veteran    TRUE       0.0         Agender           20 
173 Regular    FALSE      0.0         Male              50 
174 Veteran    TRUE       0.0         Male              17 
175 Amateur    TRUE       0.3         Male              17 
176 Pro        FALSE      0.0         Non-binary        17 
177 Veteran    FALSE      2.7         Non-binary        21 
178 Amateur    TRUE       0.4         Two-Spirited      17 
179 Veteran    TRUE       0.1         Non-binary        17 
180 Amateur    FALSE      2.1         Male              24 
181 Amateur    TRUE       0.8         Female            22 
182 Pro        FALSE      0.2         Male              17 
183 Amateur    TRUE      32.0         Male              22 
184 Pro        FALSE      1.7         Male              17 
185 Regular    FALSE      0.1         Male              18 
186 Veteran    TRUE       0.1         Female            44 
187 Amateur    TRUE       0.0         Male              17 
188 Beginner   TRUE       0.0         Female            17 
189 Amateur    FALSE      0.0         Male              17 
190 Amateur    TRUE       0.0         Male              20 
191 Amateur    TRUE       0.0         Female            17 
192 Veteran    FALSE      0.3         Male              22 
193 Amateur    FALSE      0.0         Prefer not to say 17 
194 Amateur    FALSE      2.3         Male              17

In [13]:
split <- initial_split(players, prop = 0.75, strata = subscribe)
train <- training(split)
test <- testing(split)

In [14]:
vfold <- vfold_cv(train, v = 5, strata = subscribe)

recipe <- recipe(subscribe ~ played_hours + Age, data = train) |>
  step_scale(all_predictors()) |>
  step_center(all_predictors())

knn_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = 3) |>
  set_engine("kknn") |>
  set_mode("classification")

knn_fit <- workflow() |>
  add_recipe(recipe) |>
  add_model(knn_spec) |>
  fit_resamples(resamples = vfold)|>
                  collect_metrics()

knn_fit

p_vfold <- vfold_cv(train, v = 10, strata = subscribe)

vfold_metrics <- workflow() |>
                  add_recipe(recipe) |>
                  add_model(knn_spec) |>
                  fit_resamples(resamples = p_vfold) |>
                  collect_metrics()

vfold_metrics

.metric  .estimator mean      n std_err    .config             
1 accuracy binary     0.5586371 5 0.04645523 Preprocessor1_Model1
2 roc_auc  binary     0.4940476 5 0.04679344 Preprocessor1_Model1

.metric  .estimator mean      n  std_err    .config             
1 accuracy binary     0.5862271 10 0.02813311 Preprocessor1_Model1
2 roc_auc  binary     0.5298864 10 0.04608404 Preprocessor1_Model1

In [17]:
knn_spe <- nearest_neighbor(weight_func = "rectangular",
                             neighbors = tune()) |>
  set_engine("kknn") |>
  set_mode("classification")

k_vals <- tibble(neighbors = seq(from = 1, to = 100, by = 5))

knn_results <- workflow() |>
  add_recipe(recipe) |>
  add_model(knn_spe) |>
  tune_grid(resamples = p_vfold, grid = k_vals) |>
  collect_metrics()

accuracies <- knn_results |>
  filter(.metric == "accuracy")

accuracies

neighbors .metric  .estimator mean      n  std_err     .config              
1   1        accuracy binary     0.4551282 10 0.041352745 Preprocessor1_Model01
2   6        accuracy binary     0.6338462 10 0.021128487 Preprocessor1_Model02
3  11        accuracy binary     0.5809890 10 0.041527343 Preprocessor1_Model03
4  16        accuracy binary     0.6697070 10 0.020215006 Preprocessor1_Model04
5  21        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model05
6  26        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model06
7  31        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model07
8  36        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model08
9  41        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model09
10 46        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model10
11 51        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model11
12 56        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model12
13 61        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model13
14 66        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model14
15 71        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model15
16 76        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model16
17 81        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model17
18 86        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model18
19 91        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model19
20 96        accuracy binary     0.7312088 10 0.005090213 Preprocessor1_Model20

In [18]:
best_k <- accuracies |>
        arrange(desc(mean)) |>
        head(1) |>
        pull(neighbors)
best_k

[1] 21

In [21]:
precipe <- recipe(subscribe ~ played_hours + Age, data = train) |>
  step_scale(all_predictors()) |>
  step_center(all_predictors())

knn_sp <- nearest_neighbor(weight_func = "rectangular", neighbors = best_k) |>
  set_engine("kknn") |>
  set_mode("classification")

knn_fitp <- workflow() |>
  add_recipe(precipe) |>
  add_model(knn_sp) |>
  fit(data = train)

knn_fitp

══ Workflow [trained] ══════════════════════════════════════════════════════════
Preprocessor: Recipe
Model: nearest_neighbor()

── Preprocessor ────────────────────────────────────────────────────────────────
2 Recipe Steps

• step_scale()
• step_center()

── Model ───────────────────────────────────────────────────────────────────────

Call:
kknn::train.kknn(formula = ..y ~ ., data = data, ks = min_rows(21,     data, 5), kernel = ~"rectangular")

Type of response variable: nominal
Minimal misclassification: 0.2689655
Best kernel: rectangular
Best k: 21

In [25]:
p_predictions <- predict(knn_fitp, test) |>
  bind_cols(test)

p_predictions |>
  metrics(truth = subscribe, estimate = .pred_class) |>
  filter(.metric == "accuracy")

p_predictions |> pull(subscribe) |> levels()

.metric  .estimator .estimate
1 accuracy binary     0.7346939

[1] "FALSE" "TRUE"

In [26]:
p_predictions |>
    precision(truth = subscribe, estimate = .pred_class, event_level = "second")

.metric   .estimator .estimate
1 precision binary     0.7346939

In [27]:
p_predictions |>
    recall(truth = subscribe, estimate = .pred_class, event_level="second")

.metric .estimator .estimate
1 recall  binary     1

Introduction:

provide some relevant background information on the topic so that someone unfamiliar with it will be prepared to understand the rest of your report
clearly state the question you tried to answer with your project
identify and fully describe the dataset that was used to answer the question

The gaming industry is expanding by the year, and efforts are made to study player's actions for a more targetted recruitment. To achieve this, researchers must predict what types of players are more likely to contribute to the games and determine which player characteristics and behaviours are most predictive of subscribing to a game-related newsletter, and how these features differ between various player types. This science project aims to specifically answer whether play hours and the player's age predict whether a player will subscribe to the Minecraft newsletter or not. In order to answer this, we used data obtained from a Minecraft server which contain specific player information and whether they subscribe to the newsletter or not. The dataset, named "players", portrays a list of players and their data (7 variables, 196 observations): player's experience (amateur, beginner, pro, regular, veteran), subscription status (TRUE: subscribed to newsletter, FALSE: not subscribed), hashed Email that uniquely identifies each player, game play hours, player name, player gender, and player age. 
The dataset reveals that 144 players are subscribed, and 52 are not. Play hours range from 0 to 223.1 hours, and age range from 8 to 50 years old.
